koku/api/provider/models.py

```diff
@@ -197,12 +197,187 @@ def delete(self, *args, **kwargs):
+   set infrastructure_id = %s
+  from public.api_providerinfrastructuremap m
+ where m.infrastructure_provider_id = %s::uuid
+   and p.infrastructure_id = m.id
+;
+"""  # noqa: F541
+        LOG.info(
+            "Setting the infrastructure_id to null for any provider "
+            + "records that link to the target infrastructure map records"
+        )
+        params = (None, self.pk)
+        with transaction.get_connection().cursor() as cur:
+            cur.execute(_sql, params)
+
+    def _get_linked_table_names(self, normalized_type, target_table, public_schema):
+        """Given a table name and schema name and type,
+        find the other tables that are related by foreign keys"""
+        with transaction.get_connection().cursor() as cur:
+            link_table_sql = """
+select ftn.nspname as "table_schema",
+       ft.relname as "table_name",
+       fc.attname as "column_name"
+  from pg_class t
+  join pg_constraint con
+    on con.confrelid = t.oid
+  join pg_attribute fc
+    on fc.attrelid = con.conrelid
+   and fc.attnum = any(con.conkey::int[])
+  join pg_class ft
+    on ft.oid = con.conrelid
+  join pg_namespace tn
+    on tn.oid = t.relnamespace
+  join pg_namespace ftn
+    on ftn.oid = ft.relnamespace
+ where t.relname = %(target_table)s
+   and tn.nspname = %(public_schema)s
+   and con.contype = %(fk_constraint)s
+   and ftn.nspname in (%(public_schema)s, %(search_schema)s)
+   and ft.relkind = any(%(relkind)s)
+   and not ft.relispartition
+   and (
+         ft.relname ~ %(type_fregex)s or
+         ft.relname ~ %(ocptype_fregex)s or
+         ft.relname ~ %(rpt_common_fregex)s or
+         ft.relname ~ %(api_fregex)s
+       )
+ order
+    by case when ft.relname ~ %(ui_table_sregex)s then %(ui_table_sval)s
+            when ft.relname ~ %(ocp_type_sregex)s then %(ocp_type_sval)s
+            when ft.relname ~ %(daily_summ_sregex)s then %(daily_summ_sval)s
+            when ft.relname ~ %(api_sregex)s then %(api_sval)s
+            else %(default_sval)s
+       end::int,
+       ft.relname
+;
+"""
+            params = {
+                "relkind": ["p", "r"],
+                "fk_constraint": "f",
+                "target_table": target_table,
+                "public_schema": public_schema,
+                "search_schema": self.customer.schema_name,
+                "type_fregex": f"_{normalized_type}",
+                "ocptype_fregex": f"_ocp({normalized_type}|all)",
+                "rpt_common_fregex": "^reporting_common_",
+                "api_fregex": "^api_",
+                "ui_table_sregex": f"^reporting_{normalized_type}_",
+                "ui_table_sval": 1,
+                "ocp_type_sregex": f"^reporting_ocp({normalized_type}|all)",
+                "ocp_type_sval": 2,
+                "daily_summ_sregex": "_daily_summary",
+                "daily_summ_sval": 3,
+                "api_sregex": "^api_",
+                "api_sval": 10,
+                "default_sval": 5,
+            }
```

Comment:
Many of these are constants. Can we put those directly into the sql instead of parameterizing them?
example: instead of:
```
and ft.relkind = any(%(relkind)s)
```
can we do:
```
and ft.relkind = any(["p", "r"])
```

Reply:
It just seemed safer this way if, in the future, we need to change something.
Besides, there be f-strings here for parameters.